### PR TITLE
Add support for Redfish session create, delete, and authenticate

### LINKED ISSUE #1975: redfish_info / redfish_command / redfish_config: Authentication with Session Token
### Summary

For security reasons, I am trying to authenticate with a Session Token given by the Overall Manager infrastructure (HPE OneView) into numerous managed blades/compute nodes. 

However, the only authentication method I could find with redfish_* modules part of the community.general collection is username/password.

Would it be possible to authenticate with a session key/token provided by a trusted infra component ?

Thanks

### Issue Type

Feature Idea

### Component Name

redfish_info redfish_config redfish_command

### Additional Information

Providing a session key authentication method in addition to the current username/password authentication for redfish_info, redfish_command and redfish_config modules, will increase the Ansible Redfish galaxy collection traction among current and upcoming software defined infrastructures.

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by billdodd**
##### SUMMARY

The Redfish modules currently only support basic authentication. But it is common for Redfish users to use session login authentication. This change adds two commands, `CreateSession` and `DeleteSession`, to support session creation and deletion. And it also adds an `auth_token` parameter to pass a security token to other Redfish commands for authentication (as an alternative to username and password).

Fixes #1975 

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

redfish_command.py
redfish_config.py
redfish_info.py
idrac_redfish_command.py
idrac_redfish_config.py
idrac_redfish_info.py
redfish_utils.py

##### ADDITIONAL INFORMATION

Here is an example playbook showing the use of the two new commands (`CreateSession` and `DeleteSession`) and two new parameters (`auth_token` and `session_uri`): 
```paste below
---
- hosts: myhosts
  connection: local
  name: Test session login
  gather_facts: False

  vars:
    datatype: SystemHealth

  tasks:

  - name: Define output file
    include_tasks: create_output_file.yml

  - name: Create session
    redfish_command:
      category: Sessions
      command: CreateSession
      baseuri: "{{ baseuri }}"
      username: "{{ username }}"
      password: "{{ password }}"
    register: result
  - ansible.builtin.debug:
      msg: "{{ result }}"

  - name: Getting system health report
    redfish_info:
      category: Systems
      command: GetHealthReport
      baseuri: "{{ baseuri }}"
      auth_token: "{{ result.session.token }}"
    register: health

  - name: Copy results to output file
    copy:
      content: "{{ health | to_nice_json }}"
      dest: "{{ template }}.json"

  - name: Turn LED off
    redfish_command:
      category: Chassis
      command: IndicatorLedOff
      baseuri: "{{ baseuri }}"
      auth_token: "{{ result.session.token }}"

  - name: Restart system power gracefully
    redfish_command:
      category: Systems
      command: PowerGracefulRestart
      baseuri: "{{ baseuri }}"
      auth_token: "{{ result.session.token }}"

  - name: Delete session
    redfish_command:
      category: Sessions
      command: DeleteSession
      baseuri: "{{ baseuri }}"
      auth_token: "{{ result.session.token }}"
      session_uri: "{{ result.session.uri }}"
```

The `create_output_file.yml` file referenced above is:
```
---
  - name: Define timestamp
    set_fact: timestamp="{{ lookup('pipe', 'date +%Y%m%d_%H%M%S') }}"
    run_once: true

  - name: Define file to place results
    set_fact: template={{rootdir}}/{{host}}/{{host}}_{{datatype}}_{{timestamp}}

  - name: Create dropoff directory for host
    file:
      path: "{{ rootdir }}/{{ host }}"
      state: directory
```

### CODE REVIEW

**Comment by billdodd** [comment]
Marking this WIP initially in order to get feedback from Redfish developers and operators.

**Comment by billdodd** [comment]
Here's a log snippet showing the CI error:
```
2021-03-16T20:15:53.1857359Z 00:20 ansible-galaxy [core 2.11.0b2.post0] 
2021-03-16T20:15:53.1857921Z 00:20   config file = None
2021-03-16T20:15:53.1858802Z 00:20   configured module search path = ['/home/AzDevOps_azpcontainer/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
2021-03-16T20:15:53.1859942Z 00:20   ansible python module location = /home/AzDevOps_azpcontainer/.local/lib/python3.8/site-packages/ansible
2021-03-16T20:15:53.1860514Z 00:20   ansible collection location = /__w/1
2021-03-16T20:15:53.1861414Z 00:20   executable location = /home/AzDevOps_azpcontainer/.local/bin/ansible-galaxy
2021-03-16T20:15:53.1862065Z 00:20   python version = 3.8.5 (default, Jul 28 2020, 12:59:40) [GCC 9.3.0]
2021-03-16T20:15:53.1862679Z 00:20   jinja version = 2.11.2
2021-03-16T20:15:53.1863053Z 00:20   libyaml = False
2021-03-16T20:15:53.1863446Z 00:20 No config file found; using defaults
2021-03-16T20:15:53.1868999Z 00:20 Starting galaxy collection install process
2021-03-16T20:15:53.1966977Z 00:20 Found installed collection community.general:2.3.0 at '/__w/1/ansible_collections/community/general'
2021-03-16T20:15:53.1968398Z 00:20 Found installed collection community.internal_test_tools:0.3.0 at '/__w/1/ansible_collections/community/internal_test_tools'
2021-03-16T20:15:53.1969030Z 00:20 Process install dependency map
2021-03-16T20:15:53.1986585Z 00:20 Opened /home/AzDevOps_azpcontainer/.ansible/galaxy_token
2021-03-16T20:15:55.2885471Z 00:22 ERROR! Error when getting available collection versions for ansible.posix from default (https://galaxy.ansible.com/api/) (HTTP Code: 520, Message: Origin Error Code: Unknown)
2021-03-16T20:15:55.3268465Z 00:22 + result=1
2021-03-16T20:15:55.3269665Z 00:22 + set -e
2021-03-16T20:15:55.3270337Z 00:22 + '[' 1 == 0 ']'
2021-03-16T20:15:55.3270981Z 00:22 + echo '@* -> 1'
2021-03-16T20:15:55.3271516Z 00:22 @* -> 1
2021-03-16T20:15:55.3272248Z 00:22 + echo 'Command '\''@*'\'' failed 3 times!'
2021-03-16T20:15:55.3272899Z 00:22 Command '@*' failed 3 times!
2021-03-16T20:15:55.3273568Z 00:22 + exit -1
2021-03-16T20:15:55.3359955Z ##[error]Bash exited with code '255'.
```

It doesn't seem to me like this is an error in the PR. Galaxy infrastructure problem?

**Comment by felixfontein** [comment]
@billdodd yes, unfortunately this happens a lot with Galaxy (https://github.com/ansible/galaxy/issues/2429). It is a cloudflare error, I guess cloudflare thinks that the CI is starting a DDoS attack and starts blocking... I've restarted the failed test run.

File: plugins/modules/remote_management/redfish/idrac_redfish_command.py [hunk 1]
```diff
@@ -33,15 +33,17 @@
...
     type: str
   password:
-    required: true
     description:
       - Password for authentication with OOB controller
     type: str
+  auth_token:
+    description:
+      - Security token for authentication with OOB controller
+    type: str
```
Comment (felixfontein) [inline plugins/modules/remote_management/redfish/idrac_redfish_command.py:46]:
```suggestion
    type: str
    version_added: 2.3.0
```

File: plugins/modules/remote_management/redfish/idrac_redfish_config.py [hunk 2]
```diff
@@ -36,15 +36,17 @@
...
     type: str
   password:
-    required: true
     description:
       - Password for authentication with iDRAC
     type: str
+  auth_token:
+    description:
+      - Security token for authentication with OOB controller
+    type: str
```
Comment (felixfontein) [inline plugins/modules/remote_management/redfish/idrac_redfish_config.py:49]:
```suggestion
    type: str
    version_added: 2.3.0
```

File: plugins/modules/remote_management/redfish/idrac_redfish_info.py [hunk 3]
```diff
@@ -37,15 +37,17 @@
...
     type: str
   password:
-    required: true
     description:
       - Password for authentication with iDRAC controller
     type: str
+  auth_token:
+    description:
+      - Security token for authentication with OOB controller
+    type: str
```
Comment (felixfontein) [inline plugins/modules/remote_management/redfish/idrac_redfish_info.py:50]:
```suggestion
    type: str
    version_added: 2.3.0
```

File: plugins/modules/remote_management/redfish/redfish_command.py [hunk 4]
```diff
@@ -35,15 +35,21 @@
...
     type: str
   password:
-    required: true
     description:
       - Password for authentication with OOB controller
     type: str
+  auth_token:
+    description:
+      - Security token for authentication with OOB controller
+    type: str
```
Comment (felixfontein) [inline plugins/modules/remote_management/redfish/redfish_command.py:48]:
```suggestion
    type: str
    version_added: 2.3.0
```

File: plugins/modules/remote_management/redfish/redfish_command.py [hunk 5]
```diff
@@ -35,15 +35,21 @@
...
+  session_uri:
+    description:
+      - URI of the session resource
+    type: str
```
Comment (felixfontein) [inline plugins/modules/remote_management/redfish/redfish_command.py:52]:
```suggestion
    type: str
    version_added: 2.3.0
```

File: plugins/modules/remote_management/redfish/redfish_config.py [hunk 6]
```diff
@@ -34,15 +34,17 @@
...
     type: str
   password:
-    required: true
     description:
       - Password for authentication with OOB controller
     type: str
+  auth_token:
+    description:
+      - Security token for authentication with OOB controller
+    type: str
```
Comment (felixfontein) [inline plugins/modules/remote_management/redfish/redfish_config.py:47]:
```suggestion
    type: str
    version_added: 2.3.0
```

File: plugins/modules/remote_management/redfish/redfish_info.py [hunk 7]
```diff
@@ -37,15 +37,17 @@
...
     type: str
   password:
-    required: true
     description:
       - Password for authentication with OOB controller
     type: str
+  auth_token:
+    description:
+      - Security token for authentication with OOB controller
+    type: str
```
Comment (felixfontein) [inline plugins/modules/remote_management/redfish/redfish_info.py:50]:
```suggestion
    type: str
    version_added: 2.3.0
```

**Comment by billdodd** [comment]
Removed the WIP prefix from the title after feedback from the team.

**Comment by billdodd** [comment]
ready_for_review

**Comment by mraineri** [comment]
shipit

**Comment by felixfontein** [comment]
@billdodd thanks for implementing this!
@mraineri thanks for reviewing!

It's great to see that the merge part of the bot still works :)

**Comment by felixfontein** [comment]
@billdodd looks like the automatic backport to stable-2 failed (https://github.com/ansible-collections/community.general/pull/2027/checks?check_run_id=2152265146). Would you mind creating a manual backport for the stable-2 branch, so this can get included in the 2.3.0 release? (The instructions on https://docs.ansible.com/ansible/latest/community/development_process.html#backporting-merged-prs-in-ansible-base are mostly valid for this; the main difference is that we backport new features, and that this is a different repo ;) )

**Comment by billdodd** [comment]
> Would you mind creating a manual backport for the stable-2 branch, so this can get included in the 2.3.0 release?

Will do. I'll work on that on Monday.

**Comment by billdodd** [comment]
@felixfontein PR #2053 created for stable-2 backport.

**Pull request merged**